rats-tls/src/core/cpu.c

```diff
@@ -128,3 +133,104 @@ bool is_tdguest_supported(void)
 	/* "IntelTDX    " */
 	return (sig[1] == 0x65746e49) && (sig[3] == 0x5844546c) && (sig[2] == 0x20202020);
 }
+#else
+static inline void __cpuidex(int a[4], int b, int c)
+{
+	a[0] = b;
+	a[2] = c;
+	ocall_cpuid(&a[0], &a[1], &a[2], &a[3]);
+}
+
+static bool is_legacy_oot_kernel_driver(void)
+{
+	bool retval;
+
+	ocall_is_sgx_dev(&retval, "/dev/isgx");
+
+	return retval;
+}
+
+/* Prior to DCAP 1.10 release, the DCAP OOT driver uses this legacy
+ * name.
+ */
+static bool is_dcap_1_9_oot_kernel_driver(void)
+{
+	bool retval;
+
+	ocall_is_sgx_dev(&retval, "/dev/sgx/enclave");
+
+	return retval;
+}
+
+/* Since DCAP 1.10 release, the DCAP OOT driver uses the same name
+ * as in-tree driver.
+ */
+static bool is_in_tree_kernel_driver(void)
+{
+	bool retval;
+
+	ocall_is_sgx_dev(&retval, "/dev/sgx_enclave");
+
+	return retval;
+}
+
+/* return true means in sgx1 enabled */
+static bool __is_sgx1_supported(void)
+{
+	int cpu_info[4] = { 0, 0, 0, 0 };
+
+	__cpuidex(cpu_info, SGX_CPUID, 0);
+
+	return !!(cpu_info[0] & SGX1_STRING);
+}
+
+static bool __is_sgx2_supported(void)
+{
+	int cpu_info[4] = { 0, 0, 0, 0 };
+
+	__cpuidex(cpu_info, SGX_CPUID, 0);
+
+	return !!(cpu_info[0] & SGX2_STRING);
+}
+
+bool is_sgx1_supported(void)
+{
+	if (!__is_sgx1_supported())
+		return false;
+
+	/* SGX2 using ECDSA attestation is not compatible with SGX1
+         * which uses EPID attestation.
+         */
+	if (is_sgx2_supported())
+		return false;
+
+	/* Check whether the kernel driver is accessible */
+	if (!is_legacy_oot_kernel_driver())
+		return false;
+
+	return true;
+}
+
+bool is_sgx2_supported(void)
+{
+	if (!__is_sgx2_supported())
+		return false;
+
+	/* Check whether the kernel driver is accessible */
+	if (!is_dcap_1_9_oot_kernel_driver() && !is_in_tree_kernel_driver())
+		return false;
+
+	return true;
+}
+
+/* return true means in td guest */
+bool is_tdguest_supported(void)
+{
+	uint32_t sig[4] = { 0, 0, 0, 0 };
+
+	__cpuidex(sig, TDX_CPUID, 0);
+
+	/* "IntelTDX    " */
+	return (sig[1] == 0x65746e49) && (sig[3] == 0x5844546c) && (sig[2] == 0x20202020);
+}
```

Comment:
Almost all above functions are similar with non-sgx functions. So why duplicates them?

If I were you, I will implement it in this way:
```C
# ifndef SGX
static inline void cpuid(int *eax, int *ebx, int *ecx, int *edx)
{
  // ...
}

static bool is_sgx_device(const char *dev)
{
	struct stat st;

	if (!stat(dev, &st)) {
		if ((st.st_mode & S_IFCHR) && (major(st.st_rdev) == SGX_DEVICE_MAJOR_NUM))
			return true;
	}

	return false;
}
# else
# define cpuid ocall_cpuid

static bool is_sgx_device(const char *dev)
{
	bool retval;

	ocall_is_sgx_dev(&retval, dev);

	return retval;
}
# endif

// Keep the non-sgx functions as usual

static inline void __cpuidex(int a[4], int b, int c)
{
  // ...
}

// ...
```